pkg/symbol/symbolsearcher/symbol_searcher.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package symbolsearcher
+
+import (
+	"debug/elf"
+	"errors"
+	"sort"
+	"strings"
+)
+
+type Searcher struct {
+	symbols []elf.Symbol
+}
+
+func New(syms []elf.Symbol) Searcher {
+	newSyms := make([]elf.Symbol, 0, len(syms))
+	for _, s := range syms {
+		if isFunction(s) {
+			newSyms = append(newSyms, s)
+		}
+	}
+
+	// slice stable sort to keep output consistent
+	sort.SliceStable(newSyms, func(i, j int) bool {
+		if newSyms[i].Value != newSyms[j].Value {
+			return newSyms[i].Value < newSyms[j].Value
+		}
+		return chooseBestSymbol(newSyms[i], newSyms[j])
+	})
+	return Searcher{
+		symbols: newSyms,
+	}
+}
+
+func (s Searcher) Search(addr uint64) (string, error) {
+	i := sort.Search(len(s.symbols), func(i int) bool {
+		sym := s.symbols[i]
+		return sym.Value > addr
+	})
+	if i == 0 ||
+		// addr < sym[i-1]
+		addr < s.symbols[i-1].Value {
+		return "", errors.New("failed to find symbol for address")
+	}
+
+	// sym[i-1] <= addr < sym[i]
+	i--
```

Comment:
```suggestion
		if i < 1 || i > len(lnr.symbols) {
		level.Debug(lnr.logger).Log("msg", "failed to find symbol for address", "addr", addr)
		return nil, errors.New("failed to find symbol for address")
	}
```


🤔 I prefer previous bound checks and early returns. Could we maybe keep it? And these checks additionally?

Also, it would really help to add some comments.

Reply:
yes, the old logic is better